### DISCUSSION TITLE
SQLite backend

### DISCUSSION THREAD

**Discussion by felipe19930** [akiver/cs-demo-manager#741] [Q&A]
### What problem is this solving

Simplifying installation and usage for regular users.

### Proposed solution

Having a SQLite backend would simplify greatly the installation and usage for users.

SQLite is arguably lighter (disk space wise) than PostgreSQL, and doesn't require a server service running, or setting up passwords.

**Comment by leakedby** [comment]
Yes, why not Sqlite?

**Comment by H1ghSyst3m** [comment]
SQLite would be so good, because the database can then also be saved then if you have to change or reset computer. The Database would be portable then

**Comment by felipe19930** [comment]
@akiver I see you seem to have no interest in this, but if someone else were to implement this, and then send their work for review through a pull request, would you even consider accepting it?

**Comment by jamezmoran** [comment]
Any thoughts on why postgres exactly is the default here? I'm looking at the code now and so far don't see any reasons this can't be ported to SQLite, @akiver appears to have taken care to not encumber the app in postgres specific things which is good. Just trying to understand where the sticking points are incase I (or someone else) might want to knock this out in a week, if that's even possible.

**Reply by ShiinaKin** [reply]
Agree. IMO, it's not difficult to resolve this problem, maybe just add a setting page to choose which Database he want to use, and use correspondingly driver to implement it at the low level. But I'm a Java programmer, maybe in Node do this will be difficult, otherwise I can't think of any reason why shouldn't do this.

**Comment by LouisAsanaka** [comment]
I'd be willing to give it a shot when I'm free, provided that @akiver approves of the inclusion of SQLite :)

**Comment by alirezanet** [comment]
Postgres db?! is the worst choice ever :)  !!!!!!!

**Comment by guixxx** [comment]
+1 for a lighter database backend. 

![image](https://github.com/akiver/cs-demo-manager/assets/27077338/de1039a9-c217-49ba-aa3d-77812915b752)

PostgreSQL using almost 4x more disk space than the actual app is ludicrous.

**Comment by FalkF** [comment]
@akiver I feel like with minimal input many people could end up happy long term ;)

**Comment by davimarcilio** [comment]
I gave this a shot, but with a bundled PostgreSQL instead of SQLite: #1452 (draft).

The reason for not going the SQLite route is that the schema and the queries lean on PostgreSQL quite a bit — `DISTINCT ON`, aggregate `FILTER (WHERE …)`, `percentile_cont … WITHIN GROUP`, window functions, PL/pgSQL triggers, `GENERATED ALWAYS AS … STORED`, ~25 CTEs. Shipping PostgreSQL itself leaves the 300+ Kysely queries and the 48 database migrations untouched, so it solves the installation problem without a risky rewrite of the data layer.

What it does:

- The app runs `initdb` on the first launch and starts its own cluster with `pg_ctl` on a free loopback port. No installation, no Docker, no connection form.
- `psql` stops being a dependency entirely — matches now stream through `COPY … FROM STDIN` on the existing connection instead of shelling out to `psql \copy`. That also removes the class of problems reported in #955 and #890 for people using an external server.
- Existing installations keep using their external server through a settings migration, and the mode can be switched in Settings → Database.
- About 60 MB added to the Windows build.

@akiver it's a large change and I understand if it isn't a direction you want for the project — I'd rather hear that now than after polishing it further. macOS and Linux still need to be tested, only Windows was verified so far.

**Reply by zneel** [reply]
@davimarcilio there's PGLite that can be a drop-in replacement withtout changing the queries i guess ? https://github.com/electric-sql/pglite

**Reply by davimarcilio** [reply]
Good call, and the dialect part of your point is right — PGlite is real PostgreSQL, so `DISTINCT ON`, aggregate `FILTER`, the PL/pgSQL triggers and the generated columns would all work unchanged. The blocker isn't SQL, it's the process model.

@akiver already tried it and hit that wall (from https://github.com/akiver/cs-demo-manager/pull/1452#issuecomment-5304194297):

> I tried a long time ago to use https://pglite.dev to have a local PG database. It worked, but the downside is that we can't have both the app and the CLI connected to the database at the same time (it may corrupt the FS because PGlite is single-connection).

Two things it runs into here:

1. **The app and the CLI both need the database.** PGlite runs in-process and is single-connection, so `csdm` on the command line while the app is open can't work. Fronting it with `pglite-socket` fixes that, but then you're back to running a server process, which is the thing bundling was supposed to avoid. The native cluster handles it by reusing the running postmaster through `postmaster.pid`.
2. **Match insertion is concurrency-bound.** A demo analyzed with positions writes ~600K rows to `player_positions` alone, and the insertion fires 33 concurrent `COPY … FROM STDIN` commands over a pool of 8 connections. That needs real connection concurrency.

Where it landed: #1453 (the `psql` removal) is merged, and #1454 has the bundled cluster on top of it.
